### ISSUE TITLE
[scx_lavd] Getting large pauses and stutters in-game

### ISSUE THREAD

**Issue by ChrisLane** [sched-ext/scx#234]
I read about scx_lavd on Phoronix and decided to give it a try and get some basic benchmarks.

I have observed that I am able to quite reliably reproduce stutters and pauses, sometimes several seconds long while running around my ship in Helldivers 2.

Here's the output from `sudo scx_lavd` while I was running the game with several stutters in the period (not sure on exact timestamps:
https://gist.github.com/ChrisLane/1945f66b5d8a2f36a530ca9ac8abcfa1

Please let me know if I can do anything to debug or provide more useful info for your debugging purposes.

**System**:
Distro: Arch Linux
Kernel: 6.8.7-1-cachyos
GPU: AMD Radeon RX 6800 XT (RADV NAVI21)
CPU: AMD Ryzen 9 5900X 12-Core Processor
RAM: 36GB

**Comment by Quackdoc** [comment]
Did you verify the stutters don't happen when the scheduler isn't loaded? I've been having very similar issues with all of the cachyOS kernels https://github.com/CachyOS/linux-cachyos/issues/244

**Comment by ChrisLane** [comment]
I didn't have any issues with the cachyos kernel when I took a baseline, though it was a short test.

With the frequency in which the issue was occurring with scx_lavd (often), I would have expected to notice if it was also happening on default cachyos kernel and scheduler.
I can try playing for longer on the default cachyos kernel and scheduler to fully rule that out though.

**Comment by Cifer9516** [comment]
Same thing, heavy load scenarios time out the scheduler and it just closes.
Happened in vkd3d game and during compilation.

**Comment by skygrango** [comment]
I tested compiling the rust program and playing the game at the same time, and got an scx_lavd error.

log file :  [scx_lavd_dump-2024-04-20-1.txt](https://github.com/sched-ext/scx/files/15046301/scx_lavd_dump-2024-04-20-1.txt)

**Comment by multics69** [comment]
Thank you for reporting the issues. This will really helpful to improve and tune the LAVD. The preemption logic is under development and it will be improved next cycle.

@ChrisLane -- Thank you for opening the issue and share the log!

@Cifer9516  -- If you don't mind, could you share the log and the game scenario that you played?

@skygrango  -- If you don't mind, please share the game scenario that you played?

**Comment by bunkbail** [comment]
here's a dump log of `sudo scx_lavd -s $(nproc)` on latest scx_lavd. i'm on debian, kernel is compiled by me. what i did was just, i launched scx_lavd, ran the game "infinite wealth", run around for a bit while the game is having freezes all over and tab out until i see lavd crashes. the game runs fine after lavd crashed. 
[dump.txt](https://github.com/sched-ext/scx/files/15098588/dump.txt)

**Comment by Galcian79** [comment]
I think i'm facing the same issue here. Stuttering in game and then LAVD crashes.
[journalctl.log](https://github.com/sched-ext/scx/files/15101760/journalctl.log)

**Comment by Cifer9516** [comment]
> @Cifer9516 -- If you don't mind, could you share the log and the game scenario that you played?
I launched vkd3d game through bottles. Here is full log
```
[user]# su
[user]# scx_lavd
00:24:10 [INFO] scx_lavd scheduler is initialized
00:24:10 [INFO] scx_lavd scheduler starts running.
00:24:11 [INFO] | mseq      | pid      | comm              | cpu  | vddln_ns  | elglty_ns | slice_ns   | grdy_rt   | lat_prio | lat_cri | min_lc  | avg_lc  | max_lc  | static_prio  | lat_bst | slice_bst | run_freq  | run_tm_ns | wait_freq | wake_freq | cpu_util |
00:24:11 [INFO] |         1 |     3377 | konsole           |    3 |         0 |         0 |      75000 |         0 |       20 |       0 |       0 |       0 |       0 |           20 |       0 |         0 |         1 |  12000000 |         0 |         0 |        0 | 
00:24:12 [INFO] |         2 |      344 | kworker/u16:9     |    1 |         0 |         0 |    3000000 |       198 |       17 |      41 |      32 |      39 |      43 |           20 |      -3 |         0 |       581 |      8920 |       745 |       156 |        2 | 
00:24:13 [INFO] |         3 |      345 | kworker/u16:10    |    0 |   4842000 |      4914 |     997973 |      3535 |       18 |      41 |      27 |      40 |      48 |           20 |      -2 |         0 |     12275 |      2246 |     16217 |      2604 |        4 | 
00:24:14 [INFO] |         4 |     3377 | konsole           |    1 |   1008000 |    242450 |    3000000 |     38380 |       11 |      51 |      36 |      40 |      49 |           20 |      -9 |         0 |      2009 |    401764 |     12348 |    416860 |        2 | 
00:24:15 [INFO] |         5 |      345 | kworker/u16:10    |    4 |         0 |         0 |    3000000 |       328 |       19 |      41 |      37 |      40 |      46 |           20 |      -1 |         0 |       632 |      8515 |       814 |      8314 |        0 | 
00:24:16 [INFO] |         6 |     3351 | kworker/u16:15    |    0 |   6015000 |     16171 |    3000000 |      4437 |       19 |      41 |      34 |      41 |      51 |           20 |      -1 |         0 |      4774 |      9839 |      6455 |      6594 |        1 | 
00:24:17 [INFO] |         7 |     3351 | kworker/u16:15    |    1 |         0 |         0 |    3000000 |       553 |       21 |      41 |      25 |      43 |      53 |           20 |       1 |         0 |      6129 |      6759 |      2958 |      2959 |       11 | 
00:24:18 [INFO] |         8 |      123 | kworker/u16:4     |    0 |   7500000 |     27240 |    3000000 |      1017 |       20 |      41 |      29 |      43 |      51 |           20 |       0 |         0 |       686 |      8670 |       888 |      3726 |        4 | 
00:24:19 [INFO] |         9 |     3352 | kworker/u16:16    |    0 |   7500000 |    226356 |     300000 |      6423 |       20 |      41 |      20 |      44 |      52 |           20 |       0 |         0 |       542 |     12408 |       850 |      1926 |       23 | 
00:24:20 [INFO] |        10 |     3351 | kworker/u16:15    |    0 |   9366000 |     48040 |     512971 |      1581 |       21 |      41 |      36 |      43 |      51 |           20 |       1 |         0 |       649 |     12176 |       844 |     19336 |        9 | 
00:24:21 [INFO] |        11 |     3352 | kworker/u16:16    |    0 |   7500000 |         0 |    3000000 |       884 |       20 |      41 |      28 |      43 |      51 |           20 |       0 |         0 |      1874 |      6506 |      2474 |      4656 |        4 | 
00:24:22 [INFO] |        12 |     3377 | konsole           |    1 |   2460000 |    472700 |     300000 |     33296 |       15 |      50 |      16 |      45 |      53 |           20 |      -5 |         0 |      1034 |    112245 |      9224 |    338266 |       12 | 
00:24:23 [INFO] |        13 |     3351 | kworker/u16:15    |    4 |   6015000 |     10846 |     300000 |      8067 |       19 |      47 |      17 |      48 |      55 |           20 |      -1 |         0 |     16088 |      7008 |     22418 |     64866 |       11 | 
00:24:24 [INFO] |        14 |     3352 | kworker/u16:16    |    5 |   7500000 |      4012 |     500529 |      2455 |       20 |      45 |      28 |      47 |      55 |           20 |       0 |         0 |     12236 |     17353 |     21511 |     36729 |       26 | 
00:24:25 [INFO] |        15 |     3352 | kworker/u16:16    |    0 |   7500000 |      7120 |    2218095 |      3287 |       20 |      45 |      35 |      47 |      56 |           20 |       0 |         0 |      8497 |     23819 |     11255 |     47552 |       22 | 
00:24:26 [INFO] |        16 |       17 | ksoftirqd/0       |    0 |   4842000 |         0 |    3000000 |         2 |       18 |      48 |      36 |      46 |      56 |           20 |      -2 |         0 |       395 |      1688 |       510 |    534225 |       26 | 
00:24:27 [INFO] |        17 |     5438 | Wwise Execution   |    0 |   3072000 |      1898 |    3000000 |      3611 |       16 |      46 |      38 |      47 |      55 |           16 |       0 |         0 |     38032 |     23134 |     38582 |     38684 |       35 | 
00:24:28 [INFO] |        18 |     3351 | kworker/u16:15    |    0 |   7500000 |      1311 |    1211071 |      2372 |       20 |      46 |      35 |      48 |      56 |           20 |       0 |         0 |     32223 |      4251 |     44781 |     28147 |       36 | 
00:24:29 [INFO] |        19 |     3377 | konsole           |    0 |         0 |         0 |    3000000 |       704 |       19 |      49 |      33 |      48 |      55 |           20 |      -1 |         0 |       607 |    128913 |      5936 |    222631 |       30 | 
00:24:30 [INFO] |        20 |     5322 | GameThread        |    3 |   1566000 |         0 |    3000000 |      3663 |       13 |      50 |      34 |      48 |      55 |           16 |      -3 |         0 |     75717 |      3135 |    135284 |     67433 |       23 | 
00:24:31 [INFO] |        21 |     5322 | GameThread        |    4 |         0 |         0 |    3000000 |      1178 |       14 |      50 |      39 |      48 |      56 |           16 |      -2 |         0 |    149393 |      3128 |    182771 |     74301 |       24 | 
00:24:32 [INFO] |        22 |     5322 | GameThread        |    0 |   1260000 |      5552 |     300000 |     39608 |       12 |      51 |      34 |      48 |      55 |           16 |      -4 |         0 |     97070 |      4066 |    190573 |     89980 |       24 | 
00:24:33 [INFO] |        23 |     2939 | IPC I/O Child     |    4 |   7500000 |     17174 |     907750 |      2578 |       20 |      46 |      36 |      48 |      55 |           20 |       0 |         0 |      3094 |     16852 |      3925 |    118110 |       19 | 
00:24:34 [INFO] |        24 |      347 | kworker/u16
00:24:11 [INFO] | mseq      | pid      | comm              | cpu  | vddln_ns  |  38347151 |     17282 |       340 |       26 | 
00:24:39 [INFO] |        29 |     5322 | GameThread        |    7 |   2460000 |         0 |    2861144 |       168 |       15 |      38 |      31 |      36 |      51 |           16 |      -1 |         0 |        11 |   7764760 |      7294 |       647 |       27 | 
00:24:40 [INFO] |        30 |     5322 | GameThread        |    6 |   4842000 |         0 |    2921168 |         3 |       18 |      37 |      33 |      39 |      51 |           16 |       2 |         0 |         3 |   7708214 |      4104 |       939 |       27 | 
00:24:41 [INFO] |        31 |     5322 | GameThread        |    2 |   2460000 |         0 |    2921168 |        15 |       15 |      41 |      35 |      39 |      51 |           16 |      -1 |         0 |         3 |    655484 |      3078 |      1913 |       28 | 
00:24:42 [INFO] |        32 |     5322 | GameT
00:24:11 [INFO] | mseq      | pid      | comm              | cpu  | vddln_ns               | cpu  | vddln_ns  | elglty_ns | slice_ns   | grdy_rt   | lat_prio | lat_cri | min_lc  | avg_lc  | max_lc  | static_prio  | lat_bst | slice_bst | run_freq  | run_tm_ns | wait_freq | wake_freq | cpu_util |
00:24:43 [INFO] |        33 |     5322 | GameThread        |    2 |   3858000 |         0 |    2901160 |       553 |       17 |      38 |      30 |      39 |      51 |           16 |       1 |         0 |         7 |   8137045 |       332 |     14803 |       26 | 
00:24:44 [INFO] |        34 |     5322 | GameThread        |    3 |   3858000 |         0 |    2921168 |         0 |       17 |      36 |      26 |      38 |      51 |           16 |       1 |         0 |         3 |   9002069 |       142 |       768 |       26 | 

DEBUG DUMP
================================================================================

kworker/u16:4[123] triggered exit kind 1026:
  runnable task stall (wineserver[5257] failed to run for 7.321s)

Backtrace:
  scx_watchdog_workfn+0x154/0x1e0
  process_one_work+0x178/0x340
  worker_thread+0x301/0x490
  kthread+0xe5/0x120
  ret_from_fork+0x31/0x50
  ret_from_fork_asm+0x1b/0x30

Runqueue states
---------------

CPU 0   : nr_run=6 flags=0x0 cpu_rel=0 ops_qseq=179542 pnt_seq=760207
          curr=FAsyncLoadingTh[5449] class=ext_sched_class

 *R FAsyncLoadingTh[5449] -3ms
      scx_state/flags=3/0xd dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a)
      cpus=ff


  R wineserver[5257] -7321ms
      scx_state/flags=3/0x9 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x8000000000000002
      cpus=ff

    __x64_sys_epoll_wait+0x19d/0x1e0
    do_syscall_64+0x83/0x170
    entry_SYSCALL_64_after_hwframe+0x78/0x80

  R Foreground Work[5324] -7321ms
      scx_state/flags=3/0x9 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x8000000000000002
      cpus=ff

    __x64_sys_futex_waitv+0xdd/0x160
    do_syscall_64+0x83/0x170
    entry_SYSCALL_64_after_hwframe+0x78/0x80

  R RenderThread 0[5412] -7321ms
      scx_state/flags=3/0x9 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x8000000000000002
      cpus=ff

    __x64_sys_futex_waitv+0xdd/0x160
    do_syscall_64+0x83/0x170
    entry_SYSCALL_64_after_hwframe+0x78/0x80

  R kworker/0:2[188] -7168ms
      scx_state/flags=3/0x9 dsq_flags=0x1 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x0
      cpus=01

    kthread+0xe5/0x120
    ret_from_fork+0x31/0x50
    ret_from_fork_asm+0x1b/0x30

  R kworker/0:1H[141] -6265ms
      scx_state/flags=3/0x9 dsq_flags=0x1 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x0
      cpus=01

    kthread+0xe5/0x120
    ret_from_fork+0x31/0x50
    ret_from_fork_asm+0x1b/0x30

CPU 3   : nr_run=1 flags=0x0 cpu_rel=0 ops_qseq=148464 pnt_seq=1902939
          curr=kworker/u16:4[123] class=ext_sched_class

 *R kworker/u16:4[123] +0ms
      scx_state/flags=3/0xd dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a)
      cpus=ff

    scx_ops_error_irq_workfn+0x36b/0x490
    irq_work_run_list+0x50/0x90
    irq_work_run+0x18/0x50
    __sysvec_irq_work+0x1c/0xc0
    sysvec_irq_work+0x6d/0x90
    asm_sysvec_irq_work+0x1a/0x20
    scx_watchdog_workfn+0x16c/0x1e0
    process_one_work+0x178/0x340
    worker_thread+0x301/0x490
    kthread+0xe5/0x120
    ret_from_fork+0x31/0x50
    ret_from_fork_asm+0x1b/0x30

CPU 6   : nr_run=1 flags=0x0 cpu_rel=0 ops_qseq=112716 pnt_seq=2616711
          curr=GameThread[5322] class=ext_sched_class

 *R GameThread[5322] +0ms
      scx_state/flags=3/0x5 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a)
      cpus=ff


================================================================================

Error: EXIT: runnable task stall (wineserver[5257] failed to run for 7.321s)
```

**Comment by multics69** [comment]
Thanks @bunkbail and @Galcian79 for sharing the logs.

**Comment by htejun** [comment]
```
 *R LikeADragon8.ex[3041] -3ms
      scx_state/flags=3/0x5 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a)
      cpus=00000fff


  R LikeADragon8.ex[2869] -40782ms
      scx_state/flags=3/0x9 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x8000000000000002
      cpus=00000fff
```
So, this suggests that 2869 has been sitting on the local DSQ for over 40s which should only be possible if either the previous task keeps running or tasks keep getting queued ahead of it. The latter happens only with explicit `ENQ_HEAD` or `ENQ_PREEMPT`, neither of which is currently being used by `lavd`, so the situation could be that 3041 is running constantly.

Now, a gaming thread being runnable constantly wouldn't be too surprising (maybe it's busy polling for some event?) but what should have happend is it running down its slice and then taking turns with other runnable tasks, which doesn't seem to have happened. Hmm... so, one peculiarity of `lavd` is that it assigns slice from `ops.running()`. I can imagine a scenario like the following:

1. The offending thread becomes runnable.
2. It starts running, `ops.running()` sets its slice according to `calc_time_slice()`.
3. It's running within its slice but it interacts with e.g. graphics driver and wakes up an RT thread.
4. The RT thread preempts the running task. `ops.stopping()` is invoked. Because the task hasn't exhausted its slice yet, it gets pushed at the top of the local DSQ so that it can continue its slice once the RT thread is done.
5. RT thread yields the CPU back. The task which got preempted is picked and starts running again. Go back to #2.

So, if I'm reading the code right, if the task keeps waking up the RT thread before its slice expires, the task can get its time slice refreshed indefinitely. If this is the case, the remedy may be:

a. Distinguish the first `ops.running()` after `ops.runnable()` and replenish slice only on the first transition.
b. Implement `ops.cpu_release()` to detect preemption by RT / DL threads and explicitly handle the condition.

I of course could be completely wrong, so please take it with a pinch of salt. From sched_ext core side, I've been thinking about adding the ability for BPF scheduler implementation to add scheduler specific information to the debug dump. Hopefully, that should make debugging from user reports easier.

**Comment by arighi** [comment]
> So, if I'm reading the code right, if the task keeps waking up the RT thread before its slice expires, the task can get its time slice refreshed indefinitely. If this is the case, the remedy may be:
> 
> a. Distinguish the first `ops.running()` after `ops.runnable()` and replenish slice only on the first transition. b. Implement `ops.cpu_release()` to detect preemption by RT / DL threads and explicitly handle the condition.

I was wondering if it could be helpful to have a way to easily migrate *all* tasks to the SCHED_EXT class, even RT/DL tasks (maybe excluding kthreads).

At some point I had a similar issue with rustland and RT audio tasks, where audio was performing better (less audio cracks) by moving the RT tasks to the SCHED_EXT class. I did that from user-space (via schedtool), but having a way directly in sched-ext to "disable" the other sched classes (even if it sounds a bit too extreme) could be useful, especially for debugging, since we could reliably take out of the equation RT tasks. Opinions?

**Comment by skygrango** [comment]
> @skygrango -- If you don't mind, please share the game scenario that you played? It seems that you ran the old version of scx_lavd. If possible, could you try the lastest version (git HEAD).

after I compile the latest git version, scx_lavd no longer crashes.

but when I run compilation with 100% cpu usage in the background, the game fps is still very unstable and low.

**Comment by htejun** [comment]
> I was wondering if it could be helpful to have a way to easily migrate _all_ tasks to the SCHED_EXT class, even RT/DL tasks (maybe excluding kthreads).

Yeah, maybe. It's not technically difficult to implement but it's also kinda nice to have the guarantee that RT tasks will always run no matter what happens to scx schedulers. Lemme think over it.

Thanks.

**Comment by Galcian79** [comment]
> > @skygrango -- If you don't mind, please share the game scenario that you played? It seems that you ran the old version of scx_lavd. If possible, could you try the lastest version (git HEAD).
> 
> after I compile the latest git version, scx_lavd no longer crashes.
> 
> but when I run compilation with 100% cpu usage in the background, the game fps is still very unstable and low.

I synced to git too. The issue is still there i think, but far less noticeable. Doesn't crash anymore.

**Comment by Galcian79** [comment]
Wouldn't be useful to have some tool like scx_lavd --version?

**Comment by sirlucjan** [comment]
> Wouldn't be useful to have some tool like scx_lavd --version?

If you use the “git” version, you can always check what you are using. For example, in Arch Linux:

```
lucjan at cachyos ~ 22:08:34    
❯ pacman -Qs scx-scheds-git                       
local/scx-scheds-git 0.1.8.r59.g973aded-1
    sched_ext schedulers
```

you can very easily indicate which commit I built from and then you know what version you have at your place (in this case, for the moment in which I publish a comment - the latest version).

Your idea is not bad, although I see one drawback. Suppose we give version “0.1.0” - and now the question is, how would it be bumped up? What release? Every commit/pull request? I say in advance this is not a criticism just a request to develop the idea.

**Comment by Galcian79** [comment]
> > Wouldn't be useful to have some tool like scx_lavd --version?
> 
> If you use the “git” version, you can always check what you are using. For example, in Arch Linux:
> 
> ```
> lucjan at cachyos ~ 22:08:34    
> ❯ pacman -Qs scx-scheds-git                       
> local/scx-scheds-git 0.1.8.r59.g973aded-1
>     sched_ext schedulers
> ```
> 
> you can very easily indicate which commit I built from and then you know what version you have at your place (in this case, for the moment in which I publish a comment - the latest version).
> 
> Your idea is not bad, although I see one drawback. Suppose we give version “0.1.0” - and now the question is, how would it be bumped up? What release? Every commit/pull request? I say in advance this is not a criticism just a request to develop the idea.

I see now. So all the schedulers are tied to the same version.

**Comment by sirlucjan** [comment]
I mean your idea is not bad. Only you would have to work out how to update the version. However, checking what commit you are based on is also a good solution, because then you know globally what version you are using.

**Comment by bunkbail** [comment]
Edited my previous comments, adding the commit hash version of the build. 

Here's a new dump log of the latest build (commit 973aded), now it took longer to crash lavd (around 5 mins compared to previous version where it took just around 1 min to crash). Game was still lagging and freezing throughout. 

[dump.txt](https://github.com/sched-ext/scx/files/15123361/dump.txt)

**Comment by multics69** [comment]
@htejun @arighi  After looking at the logs, the watchdog time out happened only when a task is in a local DSQ. I seems your hypothesis is correct -- an RT task keeps preempting an scx task so scx task's time slide is replenished every preemption, running indefinitely and causing the watchdog timeout error.

Here are my two suggestions:
1) I think it is worth to implement your option (a). If possible, it would be  better to replelish the remained time slice (not 20 msec).
2) In the watchdog logs, it would be worth to print out task's time slice. This will give us more hint where the problem comes from.

What do you think?

**Comment by arighi** [comment]
@multics69 I need to look better at the code (maybe @htejun or @Byte-Lab already have a better answer to this), but I think it should be possible already to replenish only the remaining time slice, implementing `ops.cpu_release()`, check whether a higher-priority task takes the CPU and restoring the remaining time slice in `ops.cpu_acquire()`...?

Maybe another approach to attack the problem could be to have multiple SCHED_EXT classes (following what Joel suggested here https://lore.kernel.org/all/CAEXW_YR02g=DetfwM98ZoveWEbGbGGfb1KAikcBeC=Pkvqf4OA@mail.gmail.com/). Having multiple SCHED_EXT classes would allow to have more options to decide which classes of tasks a sched-ext scheduler can grab, and eventually provide cmdline options to grab **all** tasks, or exclude RT tasks, etc.

But I'm just brainstorming here... this is something to think about more.

**Comment by multics69** [comment]
@arighi  -- Yes, restoring the remaining time slice is possible using cpu_release/acquire() callbacks. However, if it could be the problem in all schedulers, I think it would be better to be implemented on the framework level. Again, @htejun and @Byte-Lab are the right person to answer this. :-)

Supporting multiple sched_ext classes sounds an interesting direction. I will also think about more.

**Comment by htejun** [comment]
I don't think `lavd` needs to implement `ops.cpu_release/acquire()` to solve this problem. The problem is specific to `lavd` in that currently no other schedulers override `p->scx.slice` from `ops.running()` and that `lavd` is currently not considering the fact that a task may transition multiple times between `ops.running()` and `ops.stopping` between a pair of `ops.runnable()` and `ops.quiescent()`. Given those, here are possible solutions that come to mind:

- Have a per-task flag, e.g. `refresh_slice` which is set by `ops.runnable()` and make `ops_running()` set the slice iff the flag is set and then clear the flag.
- I suppose the reason why `p->scx.slice` is being set from `ops.running()` is because `scx_bpf_dispatch()` is too early to determine the slice. This is a bit of hole in the scx API design. Originally, there were only local DSQs so the expectation was that a task which is `scx_bpf_dispatch()`'d would get executed right away and thus that's where the slice was set. However, with user-DSQs this is no longer the case. Tasks can sit in user DSQs for arbitrary amount of time, so depending on the scheduler implementation, it may not be desirable to set the slice at the time a given task is dispatched to such user DSQ. Hopefully, this should be resolved with the newly introduced DSQ iterators (will update scx repo with support later today). Once this is in place, `lavd` should be able to walk its user DSQ, pick the target task to run next, set its slice and then consume the task from `ops.dispatch()` which would make it unnecessary to muck with the slice from `ops.running()`.

Other than the immediate `lavd` misbehavior, there are three issues raised in this thread:
1. Taking over all tasks including RT / DL. There's more to think about here. e.g. some RT kthreads are used for kernel's basic operation and stalling them can lead to system level failures. e.g. If kernel hang check timer kthread isn't scheduled in time, it will trigger it which in some systems would be configured to panic and reset the system, so this has the problem of weakening one of the important properties that SCX tries to guarantee. So, right now, I'm leaning more towards leaving them alone. I think, as a group, SCX scheduler writers would need to learn that there are other sched classes on top of us and deal with it. All the necessary interface is already there but we definitely can benefit from better documentation.
2. The ordering between kernel's fair class and SCX. This seems to be a feature which has been requested by multiple parties and there are no safety concerns. I'll think more on it and likely add this in some form.
3. Adding more information in debug dumps. Yeah, for sure. However, I'm not sure remaining slice printout would have helped much here. It would have looked like the task just has a normal slice duration as the problem is the slice being repeatedly being replenished (if the current hypothesis is correct). What may have been more useful is the total runtime since the last runnable transition. Also, I'm planning to add scheduler-specific dump information so that schedulers can dump whatever they want to dump.

**Comment by htejun** [comment]
Hmmm.... @arighi did mention excluding RT kthreads. That probably doesn't break any kernel mechanisms. I'll think more on it.

**Comment by multics69** [comment]
Thank you @htejun for the detailed explanation.

Setting the time slice at ops.running() seems to be the source of the problem as you explained.

Regarding the scx iterator, I don't think the iterator helps much here. That is because the global DSQ is already vtime-based DSQ. And, in lavd, vtime of DSQ is task' deadline, so scx_bpf_dispatch() should always return a task with the earliest deadline.

Regarding the refresh_slice, I thought a transition such that **ops.runnable()** -> **ops.running()** -> ops.stopping() -> **ops.running()** ... is possible even without RT task preemption. I thought it could happen when a task's time slice is exhausted but a task is still in a runnable state. Am I correct?

**Comment by htejun** [comment]
> Regarding the scx iterator, I don't think the iterator helps much here. That is because the global DSQ is already vtime-based DSQ. And, in lavd, vtime of DSQ is task' deadline, so scx_bpf_dispatch() should always return a task with the earliest deadline. Or we set task's time slice at ops.dispatch() using scx iterator?

Yes, that's what I meant. `ops.dispatch()` can set the slice using the iterator as it will have access to what will get consumed.

> Regarding the refresh_slice, I thought a transition such that **ops.runnable()** -> **ops.running()** -> ops.stopping() -> **ops.running()** ... is possible even without RT task preemption. I thought it could happen when a task's time slice is exhausted but a task is still in a runnable state. Am I correct?

You're right. I guess that's why you put slice setting in `ops.running()` in the first place. Would reasserting `refresh_slice` from `ops.stopping()` if it's exhausted work?

**Comment by htejun** [comment]
Or rather, how about `ops.runnable()` setting slice to 0 and `ops.running()` refreshing slice iff it's 0?

**Comment by htejun** [comment]
Oops, that won't work right now. Kernel will trigger zero slice warning.

**Comment by bunkbail** [comment]
Another update from me. I've built lavd from the latest commit (d9ea53c) and also cherry picked 2 patches from PR #247 and PR #250. Now, lavd no longer crashes and the game no longer stutters/freezes like before. In my limited test, everything is smooth as butter. Awesome work @multics69 and @htejun!

**Comment by multics69** [comment]
Thank you @bunkbail for testing the PRs!

**Comment by DasLeo** [comment]
@bunkbail on which kernel/distro did you test it?
I just want to verify if I'm having the same issue on my environment and exclude other kernel config/gamemode related issues.

**Comment by multics69** [comment]
@DasLeo Did you manually apply PR #247 and #250? They are not merged yet.

**Comment by bunkbail** [comment]
> @bunkbail on which kernel/distro did you test it? I just want to verify if I'm having the same issue on my environment and exclude other kernel config/gamemode related issues.

I build my own kernel but it's basically linux-cachyos on AUR, replicated and adapted for devuan ceres (debian sid + openrc) since that's what I use.

**Comment by DasLeo** [comment]
Sorry for being late to the party, just after the patches have been merged. 🙂
I wanted to first figure out that everything is working on my current kernel before I start to test `sched_ext` again.

Tested 0.1.9 on Apex Legends, where I saw the most stutters and it's performing way better now, however I still get constant micro-stutters, that I don't see without attaching an external scheduler and using plain BORE+EEVDF in Cachy Kernel.

**Comment by multics69** [comment]
Thank you @DasLeo for testing scx_lavd! I am glad to hear that the patches somewhat mitigate the problem.

For Apex Legends, could you please run `sudo scx_lavd -s $(nproc)` and share the log with your system info (CPU model, etc)?

I am working on improving the preemption in scx_lavd. Your logs will help a lot in improving the preemption logic.

**Comment by Galcian79** [comment]
Apex Legends is FTP i could vive a try too.

**Comment by DasLeo** [comment]
> Apex Legends is FTP i could give it a try too.

If you test it, make sure to enable the shader messages `DXVK_HUD` as they also produce lots of stutters, when not yet finished.
Also unlock FPS with `%command% +fps_max unlimited` argument.

It's already enough to work in the Firing Range, no need to play online multiplayer.

**Comment by Galcian79** [comment]
> > Apex Legends is FTP i could give it a try too.
> 
> If you test it, make sure to enable the shader messages `DXVK_HUD` as they also produce lots of stutters, when not yet finished. 

Even with GPL?

**Comment by DasLeo** [comment]
> > > Apex Legends is FTP i could give it a try too.
> > 
> > 
> > If you test it, make sure to enable the shader messages `DXVK_HUD` as they also produce lots of stutters, when not yet finished.
> 
> Even with GPL?

@Galcian79 if you mean [this](https://github.com/doitsujin/dxvk#graphics-pipeline-library), it's actually as stated in the docs. Shader compilation starts as soon as the mode (map) is selected in the menu, but it takes a while and keeps running during the game phase.
Apex Shaders can go up to 2GB at least in my `$HOME/.cache/nvidia/GLCache/`.
Also had to use `__GL_SHADER_DISK_CACHE_SKIP_CLEANUP=1` to prevent NVIDIA drivers from cleaning them up after `256MB` or something.

If the caching is done, the game runs butter smooth but with a slight difference when enabling `scx_lavd`.

**Comment by LethalManBoob** [comment]
I will add here that CS2 main menu is like 10fps with lavd, runs fine without

**Comment by Galcian79** [comment]
I started the game with fps unlocked, all graphic settings set to max. I played on the training ground waiting for dxvk to compile the shaders. Didn't notice any stutter.
Then i started lavd and took the logs while playing a 10 mins match. I didn't notice any stutter either.
(BTW, we won :1st_place_medal: )
[inxi.log](https://github.com/sched-ext/scx/files/15168677/inxi.log)
[1172470_stamps.log](https://github.com/sched-ext/scx/files/15168678/1172470_stamps.log)

**Comment by Galcian79** [comment]
> > I started the game with fps unlocked, all graphic settings set to max. I played on the training ground waiting for dxvk to compile the shaders. Didn't notice any stutter. Then i started lavd and took the logs while playing a 10 mins match. I didn't notice any stutter either. (BTW, we won 🥇 ) [inxi.log](https://github.com/sched-ext/scx/files/15168677/inxi.log) [1172470_stamps.log](https://github.com/sched-ext/scx/files/15168678/1172470_stamps.log)
> 
> I dont know what your specs are but i run nvidia arch

That's why i inlculded my inxi.

**Comment by LethalManBoob** [comment]
> > > I started the game with fps unlocked, all graphic settings set to max. I played on the training ground waiting for dxvk to compile the shaders. Didn't notice any stutter. Then i started lavd and took the logs while playing a 10 mins match. I didn't notice any stutter either. (BTW, we won 🥇 ) [inxi.log](https://github.com/sched-ext/scx/files/15168677/inxi.log) [1172470_stamps.log](https://github.com/sched-ext/scx/files/15168678/1172470_stamps.log)
> > 
> > 
> > I dont know what your specs are but i run nvidia arch
> 
> That's why i included my inxi.

it doesnt show your specs...

**Comment by Galcian79** [comment]
> > > > I started the game with fps unlocked, all graphic settings set to max. I played on the training ground waiting for dxvk to compile the shaders. Didn't notice any stutter. Then i started lavd and took the logs while playing a 10 mins match. I didn't notice any stutter either. (BTW, we won 🥇 ) [inxi.log](https://github.com/sched-ext/scx/files/15168677/inxi.log) [1172470_stamps.log](https://github.com/sched-ext/scx/files/15168678/1172470_stamps.log)
> > > 
> > > 
> > > I dont know what your specs are but i run nvidia arch
> > 
> > 
> > That's why i included my inxi.
> 
> it doesnt show your specs...

Which specs are we talking about?

**Comment by LethalManBoob** [comment]
![image](https://github.com/sched-ext/scx/assets/30273211/db7cbb8f-b887-4d2b-ae33-738a846aa7f9)
didnt know these two links were seperate

**Comment by Galcian79** [comment]
> ![image](https://private-user-images.githubusercontent.com/30273211/327080356-db7cbb8f-b887-4d2b-ae33-738a846aa7f9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ1NjIzODYsIm5iZiI6MTcxNDU2MjA4NiwicGF0aCI6Ii8zMDI3MzIxMS8zMjcwODAzNTYtZGI3Y2JiOGYtYjg4Ny00ZDJiLWFlMzMtNzM4YTg0NmFhN2Y5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAxVDExMTQ0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlNGFhNGIyMjE4MmIwNDlhNmMxYmY3MjA4ZDYwYTk5YzVmZGYwZjgzNGM1OWE5YmE3OTQ3N2E4M2ZjOGVkZDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TKMzYl9t__jAmrC_ThJdFjr1vJA7ZMqL54BCgzkd5Xs) didnt know these two links were seperate

Yeah my fault, i was editing on my phone and thought they showed on separate lines.

**Comment by Galcian79** [comment]
Fixed my op.

**Comment by Galcian79** [comment]
> Hey folks,
> 
> I ran a test again now to attach my [inxi.log](https://github.com/sched-ext/scx/files/15180706/inxi.log) + [scx_lavd-01-05-2024.log](https://github.com/sched-ext/scx/files/15180709/scx_lavd-01-05-2024.log)
> 
> Some facts:
> 
> * I was standing still in Apex Legends to compare fps -> dropped from 170 to 145 after starting `scx_lavd`
> * Constant Microstutter (while moving the mouse) as soon as `scx_lavd` is started, but butter-smooth on plain BORE+EEVDS
> * Running Arch, X11, KDE without Composition, Nvidia 1080 on proprietary driver 550.76
> * Find the rest in the [inxi.log](https://github.com/sched-ext/scx/files/15180706/inxi.log)
> 
> Hope this helps, otherwise just ping me.

I guess it doesn't like the llvm compiler then.

**Comment by Galcian79** [comment]
Here is the log of another game. I just found logs cannot run for more than 3hrs.

[1658280_stamps.tar.gz](https://github.com/sched-ext/scx/files/15182544/1658280_stamps.tar.gz)

**Comment by multics69** [comment]
Thanks @DasLeo and @Galcian79 for narrowing down the problem.
I will take a close look and improve it in the next version (hopefully).

**Comment by multics69** [comment]
@all -- Could you please test if the problem still exists? PR #274 should fix the problem.

**Comment by ptr1337** [comment]
> @ALL -- Could you please test if the problem still exists? PR #274 should fix the problem.

FYI, please test on the 6.9 linux-cachyos-rc kernel, since LAVD requires cpufreq support which is not included in the cachyos 6.8 Kernel currently.
Ive updated the scx-scheds-git package with the above mentioned patch.

**Comment by bunkbail** [comment]
I wanted to test the latest lavd but I only have kernel 6.8.9 and latest lavd refuses to launch there. I tried reverting commit a24e1d7 and build, but it still refuses to launch where pretty much other scx schedulers are running just fine. I'll wait until 6.9 hits stable then I'll be able to test lavd.

**Comment by Galcian79** [comment]
[1604030_lavd.tar.gz](https://github.com/sched-ext/scx/files/15279942/1604030_lavd.tar.gz)

**Comment by multics69** [comment]
@Galcian79 Thank you for sharing your log. Did you notice any stuttering? Is the problem getting worse or better?

**Comment by Galcian79** [comment]
> @Galcian79 Thank you for sharing your log. Did you notice any stuttering? Is the problem getting worse or better?

I haven't noticed any stuttering.

**Comment by DasLeo** [comment]
Tried to build Cachy 6.9-rc7 (I guess) yesterday but I got BTF errors in the final vmlinux step, need to figure this out to start further testing.

**Comment by DasLeo** [comment]
I was testing for a while today on 6.7-rc7 and hadn't had any issues nor micro-stutter.
FPS stays the same with and without the scheduler attached.

Great work folks! 👍👍👍

Here are my logs if you still wanna have a look:
- [inxi-12-05-2024.log](https://github.com/sched-ext/scx/files/15287840/inxi-12-05-2024.log)
- [scx_lavd-12-05-2024.log](https://github.com/sched-ext/scx/files/15287842/scx_lavd-12-05-2024.log)

Will use the `scx_lavd` from now on. ✌️

**Comment by bunkbail** [comment]
Same here. I'm on the stable 6.9 kernel, no micro-stutter during gaming and desktop responsiveness under load is also good. Great work @multics69!

**Comment by ChrisLane** [comment]
Compiled at fa1c146cad9bea33acfb2e1c2d65fda50382586e and tested with the same setup but 6.9.0-rc7-5-cachyos-rc and didn't experience any of the severe stuttering that I was having before :tada: 

Closing this issue.
